rationalize STATIC vs. None

avikchaudhuri opened the pull request:
Summary:
A bit of refactoring to prepare to remove `None` as a way to specify static dimensions in dynamic shapes, given we already have `Dim.STATIC` for the same purpose. We will now warn whenever this happens. However no tests were modified because problematic uses of `None` still need to behave as they do today, until we are ready to remove support. It should be easy to port tests by replacing the warning function to raise instead.

Note that other uses of `None`, such as for entire values (tensor or non-tensor) remain as is. Moving forward this should be the only purpose of `None` (at least externally).

Finally, there's a bit of confusion in our representation now because `AUTO` also internally transforms to `None`. Renamed dynamic_shapes to transformed_dynamic_shapes where this happens. Overall the two forms (pre and post transformation) have different properties so should probably not be represented in the same format in the future.

Test Plan: existing

Differential Revision: D62040729

pianpwk commented on torch/export/_trace.py:
```diff
@@ -1691,7 +1692,7 @@ def _produce_guards_callback(gm):
             fake_mode=fake_mode,
             gm=gm,
-            dynamic_shapes=dynamic_shapes,
+            dynamic_shapes=transformed_dynamic_shapes,
```
Hmm, what's the intention for returning this and feeding it to produce_guards()?

avikchaudhuri replied:
I want the `dynamic_shapes` that the user provides to be desugared early to a standard format that all internal processing uses. This way we know what the invariants are (e.g., one may have hints, the other may not). Ideally we would use the type checker to enforce invariants, but we are not there yet.

pianpwk commented on torch/export/dynamic_shapes.py:
```diff
@@ -1046,8 +1069,12 @@ def _get_dim_name_mapping(
         is_leaf=lambda x: isinstance(x, _Dim),
     )[0]:
-        if isinstance(dim, (int, _DimHint)) or dim is None:
+        if dim is None:
```
do we not warn on tensor-level Nones?

avikchaudhuri replied:
So the invariant I'm shooting for here is that things corresponding to input tree leaves can be `None` (both tensors and things like constants), but tensor dimensions cannot (because they can be STATIC).

pianpwk reviewed (approved):
generally looks good! Just the 2 questions.

avikchaudhuri committed:
rationalize STATIC vs. None (#134877)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/134877

A bit of refactoring to prepare to remove `None` as a way to specify static dimensions in dynamic shapes, given we already have `Dim.STATIC` for the same purpose. We will now warn whenever this happens. However no tests were modified because problematic uses of `None` still need to behave as they do today, until we are ready to remove support. It should be easy to port tests by replacing the warning function to raise instead.

Note that other uses of `None`, such as for entire values (tensor or non-tensor) remain as is. Moving forward this should be the only purpose of `None` (at least externally).

Finally, there's a bit of confusion in our representation now because `AUTO` also internally transforms to `None`. Renamed dynamic_shapes to transformed_dynamic_shapes where this happens. Overall the two forms (pre and post transformation) have different properties so should probably not be represented in the same format in the future.

Test Plan: existing

Reviewed By: pianpwk

Differential Revision: D62040729
